src/main/java/ome/codecs/JPEGCodec.java

```diff
@@ -138,6 +161,58 @@ public byte[] compress(byte[] data, CodecOptions options)
     return out.toByteArray();
   }
 
+  /**
+   * Create an IIOImage object with disabled chroma subsampling.
+   */
+  private IIOImage createIIOImageNoChromaSubsampling(ImageWriter jpgWriter, BufferedImage img, ImageWriteParam jpgWriteParam)
+    throws CodecException {
+    // Disable JPEG chroma subsampling
+    // http://svn.apache.org/repos/asf/shindig/trunk/java/gadgets/src/main/java/org/apache/shindig/gadgets/rewrite/image/BaseOptimizer.java
+    // http://svn.apache.org/repos/asf/shindig/trunk/java/gadgets/src/main/java/org/apache/shindig/gadgets/rewrite/image/JpegImageUtils.java
+    // Peter Haub, June. 2023
+    try {
+      IIOMetadata metadata = jpgWriter.getDefaultImageMetadata(new ImageTypeSpecifier(img.getColorModel(), img.getSampleModel()), jpgWriteParam);
+      Node rootNode = metadata!=null ? metadata.getAsTree("javax_imageio_jpeg_image_1.0") : null;
+      boolean metadataUpdated = false;
+      // The top level root node has two children, out of which the second one will
+      // contain all the information related to image markers.
+      if (rootNode!=null && rootNode.getLastChild() != null) {
+        Node markerNode = rootNode.getLastChild();
+        NodeList markers = markerNode.getChildNodes();
+        // Search for 'SOF' marker where subsampling information is stored.
+        for (int i = 0; i < markers.getLength(); i++) {
+          Node node = markers.item(i);
+          // 'SOF' marker can have
+          //   1 child node if the color representation is greyscale,
+          //   3 child nodes if the color representation is YCbCr, and
+          //   4 child nodes if the color representation is YCMK.
+          // This subsampling applies only to YCbCr.
+          if (node.getNodeName().equalsIgnoreCase("sof") && node.hasChildNodes() && node.getChildNodes().getLength() == 3) {
```

Comment:
I have not so far been able to create an example where the `SOF` node has 3 child nodes. I believe this is because `img` cannot have the relevant ColorSpace/ColorModel (see https://github.com/ome/ome-codecs/blob/master/src/main/java/ome/codecs/gui/AWTImageTools.java#L1077 and https://github.com/ome/ome-codecs/blob/master/src/main/java/ome/codecs/gui/AWTImageTools.java#L1101), but perhaps I misunderstand.

Note I tested with https://github.com/ome/bioformats/pull/4057 and the snippet therein. Is there an example that shows this code working as intended?

Reply:
**I)**
Basically, your test snippet works.

The problem is that the setting of the option `disableChromaSubsampling` is lost on the way from the snippet to the `JPEGCodec.java`.

Namely in the constructor
[public CodecOptions(CodecOptions options)](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-api/src/loci/formats/codec/CodecOptions.java#L126)
and in the function
[private static void copyOptions(CodecOptions src, ome.codecs.CodecOptions dest)](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-bsd/src/loci/formats/codec/WrappedCodec.java#L139C2-L139C82)

In both places the code must be supplemented as follows:
`dst.disableChromaSubsampling = src.disableChromaSubsampling;`
(of course also with corresponding declaration of the variable disableChromaSubsampling).

Alternatively, you can also temporarely hardcode the option disableChromaSubsampling in the JPEGCodec.java
`disableChromaSubsampling = true / false;`

**II)**
In the structured fake data (where R = G = B) the effect of the disabled chroma subsampling is not huge.
Therefore I have use the image `lena-std.tif` for my test (available in the [ImageJ example images zip archive](https://imagej.nih.gov/ij/download/sample-images.zip)).
(Of course you can use any other RGB tif image.)

I used a [modified snipped](https://github.com/ome/bioformats/pull/4057#issuecomment-1645068731) for the test.

The test shows that the resulting images (with / without chroma subsampling) are different and that the image saved **with** `disableChromaSubsampling` is closer to the original, as expected.

_(The resulting images can be opened correctly in ImageJ and QuPath. But it seems that Windows Preview and other viewers misinterpret the images as YCbCr. I could not find the necessary settings to turn this off. However, this is not relevant for the test.)_

**III)**
Apart from the basic functionality of the option `disableChromaSubsampling`, I was surprised by the fact that Tif images saved with JPEG compression (at high quality settings) are larger than the original uncompressed image.
Apparently, the image is saved line by line, which does not allow the full potential of JPEG compression to be realized. I have not investigated this circumstance further. But it should be considered to **localize** the cause of **this behavior**, because in this form JPEG compression is not really helpful.

_At this point, perhaps a word about the application:
WSI images - which are to a large extent RGB transmitted light images - are still saved by various manufacturers as Tif with JPEG compression. From my point of view it would be helpful if this format could also be written - regardless of the disadvantageous JPEG artifacts. It is precisely to minimize these artifacts that switching off chroma subsampling serves and to be able to create Tif images with JPEG compression with high quality._

**IV)**
As can be seen from the tests, the color representation does not have to be explicitly YCbCr.
The listed node conditions are also fulfilled with RGB representation.

Reply:
Note:
" ... the image is saved line by line, which does not allow the full potential of JPEG compression to be realized. ..."

In order to investigate this issue further, I have temporarily modified the function `writeImage` in `loci.formats.tiff.TiffSaver.java.`

I have temporarily inserted 
```
	if (compression.getCodecName().equals("JPEG"))      
	   ifd.putIFDValue(IFD.ROWS_PER_STRIP, new long[] {8});
```
directly after the [line](https://github.com/ome/bioformats/blob/3d4f0ca079a97070cbb647333e536f0f164d91a1/components/formats-bsd/src/loci/formats/tiff/TiffSaver.java#L313)
`	compression = ifd.getCompression();
`

This modification significantly reduces the size of the Jpeg compressed file.
In my test (with jpegquality = 0.75 and disableChromaSubsampling = false) the file size reduced from **585 kB** to **83 kB**.